kedro-datasets/kedro_datasets/databricks/managed_table_dataset.py

```diff
@@ -0,0 +1,433 @@
+    _VALID_WRITE_MODES = ["overwrite", "upsert", "append"]
+    _VALID_DATAFRAME_TYPES = ["spark", "pandas"]
+    database: str
+    catalog: str
+    table: str
+    write_mode: str
+    dataframe_type: str
+    primary_key: Optional[str]
+    owner_group: str
+    partition_columns: Union[str, List[str]]
+    json_schema: StructType
+
+    def __post_init__(self):
+        """Run validation methods if declared.
+        The validation method can be a simple check
+        that raises DataSetError.
+        The validation is performed by calling a function named:
+            `validate_<field_name>(self, value) -> raises DataSetError`
+        """
+        for name in self.__dataclass_fields__.keys():  # pylint: disable=no-member
+            method = getattr(self, f"validate_{name}", None)
+            if method:
+                method()
+
+    def validate_table(self):
+        """validates table name
+
+        Raises:
+            DataSetError:
+        """
+        if not re.fullmatch(self._NAMING_REGEX, self.table):
+            raise DataSetError("table does not conform to naming")
+
+    def validate_database(self):
+        """validates database name
+
+        Raises:
+            DataSetError: If the table name does not conform to naming constraints.
+        """
+        if self.database:
+            if not re.fullmatch(self._NAMING_REGEX, self.database):
+                raise DataSetError("database does not conform to naming")
+
+    def validate_catalog(self):
+        """validates catalog name
+
+        Raises:
+            DataSetError: If the catalog name does not conform to naming constraints.
+        """
+        if self.catalog:
+            if not re.fullmatch(self._NAMING_REGEX, self.catalog):
+                raise DataSetError("catalog does not conform to naming")
+
+    def validate_write_mode(self):
+        """validates the write mode
+
+        Raises:
+            DataSetError: If an invalid `write_mode` is passed.
+        """
+        if self.write_mode not in self._VALID_WRITE_MODES:
+            valid_modes = ", ".join(self._VALID_WRITE_MODES)
+            raise DataSetError(
+                f"Invalid `write_mode` provided: {self.write_mode}. "
+                f"`write_mode` must be one of: {valid_modes}"
+            )
+
+    def validate_dataframe_type(self):
+        """validates the dataframe type
+
+        Raises:
+            DataSetError: If an invalid `dataframe_type` is passed
+        """
+        if self.dataframe_type not in self._VALID_DATAFRAME_TYPES:
+            valid_types = ", ".join(self._VALID_DATAFRAME_TYPES)
+            raise DataSetError(f"`dataframe_type` must be one of {valid_types}")
+
+    def validate_primary_key(self):
+        """validates the primary key of the table
+
+        Raises:
+            DataSetError: If no `primary_key` is specified.
+        """
+        if self.primary_key is None or len(self.primary_key) == 0:
+            if self.write_mode == "upsert":
```

Comment:
Why only in upsert mode?

Reply:
The upsert operation works by joining on the table using its primary key. The other write modes (`append` and `overwrite`) don't need a primary key to work, as they don't perform this join.